core/src/main/java/jenkins/util/TreeString.java

```diff
@@ -124,9 +125,7 @@ public boolean equals(final Object rhs) {
     public int hashCode() {
         int h = parent == null ? 0 : parent.hashCode();
 
-        for (char c : label) {
-            h = 31 * h + c;
-        }
+        h = 31 * h + ArrayUtils.hashCode(label);
```

Comment:
This change looks wrong to me. Previously, `h` would be multiplied by 31 for every character in `label`. Now it's multiplied by 31 exactly once, regardless of the `label` length.

Reply:
This was per @jvz in https://github.com/jenkinsci/jenkins/pull/3913#discussion_r259948125

I can create a PR to switch it back to what I was originally proposing.

Reply:
A good practice is to leave hash combining to the hash function, normally by passing the seed as an argument. I think calling `ArrayUtils.deepHashCode()` would work on the combination of the precalculated parent hash and the label.

Reply:
My goal in this PR (and the series of related PRs before and after) is to have 0 changes which impacted code behavior. I'm happy to make a PR to revert this portion of this commit. But if people want to write an improved algorithm that behaves differently, I respectfully request that they do it using their knowledge.

@proski: Do you plan to write a PR to improve this code? If not, I'll just write a PR to revert this hunk.

Reply:
Please just revert that change.

Reply:
#3930

Reply:
deepHashCode would make more sense, yes. My mistake.